### DISCUSSION TITLE
Trouble getting cluster working in docker-compose

### DISCUSSION THREAD

**Discussion by tjosephs** [Grokzen/docker-redis-cluster#129] [Q&A] [answered]
I'm having a bit of trouble getting the cluster up and running. I'm not sure if this is due to my inexperience with Redis itself or something else, but I can't seem to find good answers to my specific issues anywhere, so I figured I'd ask here and see if the problem is perhaps some way that I'm misusing the image.

I'm using IORedis to try to connect my node application to the `redis-cluster` image. Here is the `docker-compose` entry that I'm using (I'm on a Mac, so I set `IP: 0.0.0.0` per the instructions in the README)

```
  redis:
    image: grokzen/redis-cluster:latest
    environment:
      IP: '0.0.0.0'
    ports:
      - '7000-7005:7000-7005'
      - '5000-5010:5000-5010'
```

Here's my TypeScript code that's trying to connect to the cluster:

 ```
import { Cluster } from 'ioredis';

const cluster = new Cluster(
  [
    {
      host: process.env.REDIS_HOST,
      port: 7000,
    },
  ],
  {
    clusterRetryStrategy: (times) => null,
  },
);
```

And here's the output that I'm getting from my application when it tries to connect to the cluster:

```
app_backend | 2021-06-23T04:33:06.414Z ioredis:cluster status: [empty] -> connecting
app_backend | Listening on port 8000
app_backend | 2021-06-23T04:33:06.436Z ioredis:cluster resolved hostname redis to IP 192.168.48.3
app_backend | 2021-06-23T04:33:06.440Z ioredis:cluster:connectionPool Reset with [ { host: '192.168.48.3', port: 7000 } ]
app_backend | 2021-06-23T04:33:06.440Z ioredis:cluster:connectionPool Connecting to 192.168.48.3:7000 as master
app_backend | 2021-06-23T04:33:06.442Z ioredis:redis status[192.168.48.3:7000]: [empty] -> wait
app_backend | 2021-06-23T04:33:06.443Z ioredis:cluster getting slot cache from 192.168.48.3:7000
app_backend | 2021-06-23T04:33:06.444Z ioredis:redis status[192.168.48.3:7000 (ioredis-cluster(refresher))]: [empty] -> wait
app_backend | 2021-06-23T04:33:06.446Z ioredis:redis status[192.168.48.3:7000 (ioredis-cluster(refresher))]: wait -> connecting
app_backend | 2021-06-23T04:33:06.447Z ioredis:redis queue command[192.168.48.3:7000 (ioredis-cluster(refresher))]: 0 -> cluster([ 'slots' ])
app_backend | 2021-06-23T04:33:06.448Z ioredis:cluster:subscriber selected a subscriber 192.168.48.3:7000
app_backend | 2021-06-23T04:33:06.448Z ioredis:redis status[192.168.48.3:7000 (ioredis-cluster(subscriber))]: [empty] -> wait
app_backend | 2021-06-23T04:33:06.449Z ioredis:cluster:subscriber started
app_backend | 2021-06-23T04:33:06.455Z ioredis:redis status[192.168.48.3:7000 (ioredis-cluster(refresher))]: connecting -> connect
app_backend | 2021-06-23T04:33:06.455Z ioredis:redis status[192.168.48.3:7000 (ioredis-cluster(refresher))]: connect -> ready
app_backend | 2021-06-23T04:33:06.456Z ioredis:connection set the connection name [ioredis-cluster(refresher)]
app_backend | 2021-06-23T04:33:06.456Z ioredis:redis write command[192.168.48.3:7000 (ioredis-cluster(refresher))]: 0 -> client([ 'setname', 'ioredis-cluster(refresher)' ])
app_backend | 2021-06-23T04:33:06.458Z ioredis:connection send 1 commands in offline queue
app_backend | 2021-06-23T04:33:06.458Z ioredis:redis write command[192.168.48.3:7000 (ioredis-cluster(refresher))]: 0 -> cluster([ 'slots' ])
app_backend | 2021-06-23T04:33:06.465Z ioredis:cluster cluster slots result count: 3
app_backend | 2021-06-23T04:33:06.466Z ioredis:cluster cluster slots result [0]: slots 0~5460 served by [ '127.0.0.1:7000', '127.0.0.1:7004' ]
app_backend | 2021-06-23T04:33:06.467Z ioredis:cluster cluster slots result [1]: slots 5461~10922 served by [ '127.0.0.1:7001', '127.0.0.1:7005' ]
app_backend | 2021-06-23T04:33:06.468Z ioredis:cluster cluster slots result [2]: slots 10923~16383 served by [ '127.0.0.1:7002', '127.0.0.1:7003' ]
app_backend | 2021-06-23T04:33:06.512Z ioredis:cluster:connectionPool Reset with [
app_backend |   { host: '127.0.0.1', port: 7000, readOnly: false },
app_backend |   { host: '127.0.0.1', port: 7004, readOnly: true },
app_backend |   { host: '127.0.0.1', port: 7001, readOnly: false },
app_backend |   { host: '127.0.0.1', port: 7005, readOnly: true },
app_backend |   { host: '127.0.0.1', port: 7002, readOnly: false },
app_backend |   { host: '127.0.0.1', port: 7003, readOnly: true }
app_backend | ]
app_backend | 2021-06-23T04:33:06.513Z ioredis:cluster:connectionPool Disconnect 192.168.48.3:7000 because the node does not hold any slot
app_backend | 2021-06-23T04:33:06.513Z ioredis:redis status[192.168.48.3:7000]: wait -> close
app_backend | 2021-06-23T04:33:06.514Z ioredis:connection skip reconnecting since the connection is manually closed.
app_backend | 2021-06-23T04:33:06.514Z ioredis:redis status[192.168.48.3:7000]: close -> end
app_backend | 2021-06-23T04:33:06.514Z ioredis:cluster:connectionPool Remove 192.168.48.3:7000 from the pool
app_backend | 2021-06-23T04:33:06.514Z ioredis:cluster:connectionPool Connecting to 127.0.0.1:7000 as master
app_backend | 2021-06-23T04:33:06.515Z ioredis:redis status[127.0.0.1:7000]: [empty] -> wait
app_backend | 2021-06-23T04:33:06.515Z ioredis:cluster:connectionPool Connecting to 127.0.0.1:7004 as slave
app_backend | 2021-06-23T04:33:06.515Z ioredis:redis status[127.0.0.1:7004]: [empty] -> wait
app_backend | 2021-06-23T04:33:06.515Z ioredis:cluster:connectionPool Connecting to 127.0.0.1:7001 as master
app_backend | 2021-06-23T04:33:06.516Z ioredis:redis status[127.0.0.1:7001]: [empty] -> wait
app_backend | 2021-06-23T04:33:06.516Z ioredis:cluster:connectionPool Connecting to 127.0.0.1:7005 as slave
app_backend | 2021-06-23T04:33:06.516Z ioredis:redis status[127.0.0.1:7005]: [empty] -> wait
app_backend | 2021-06-23T04:33:06.516Z ioredis:cluster:connectionPool Connecting to 127.0.0.1:7002 as master
app_backend | 2021-06-23T04:33:06.517Z ioredis:redis status[127.0.0.1:7002]: [empty] -> wait
app_backend | 2021-06-23T04:33:06.517Z ioredis:cluster:connectionPool Connecting to 127.0.0.1:7003 as slave
app_backend | 2021-06-23T04:33:06.517Z ioredis:redis status[127.0.0.1:7003]: [empty] -> wait
app_backend | 2021-06-23T04:33:06.518Z ioredis:cluster status: connecting -> connect
app_backend | 2021-06-23T04:33:06.519Z ioredis:redis status[127.0.0.1:7001]: wait -> connecting
app_backend | 2021-06-23T04:33:06.519Z ioredis:redis queue command[127.0.0.1:7001]: 0 -> cluster([ 'info' ])
app_backend | 2021-06-23T04:33:06.520Z ioredis:cluster:subscriber subscriber has left, selecting a new one...
app_backend | 2021-06-23T04:33:06.521Z ioredis:redis status[192.168.48.3:7000 (ioredis-cluster(subscriber))]: wait -> close
app_backend | 2021-06-23T04:33:06.521Z ioredis:connection skip reconnecting since the connection is manually closed.
app_backend | 2021-06-23T04:33:06.521Z ioredis:redis status[192.168.48.3:7000 (ioredis-cluster(subscriber))]: close -> end
app_backend | 2021-06-23T04:33:06.521Z ioredis:cluster:subscriber selected a subscriber 127.0.0.1:7002
app_backend | 2021-06-23T04:33:06.521Z ioredis:redis status[127.0.0.1:7002 (ioredis-cluster(subscriber))]: [empty] -> wait
app_backend | 2021-06-23T04:33:06.525Z ioredis:connection error: Error: connect ECONNREFUSED 127.0.0.1:7001
app_backend |     at TCPConnectWrap.afterConnect [as oncomplete] (net.js:1148:16) {
app_backend |   errno: -111,
app_backend |   code: 'ECONNREFUSED',
app_backend |   syscall: 'connect',
app_backend |   address: '127.0.0.1',
app_backend |   port: 7001
app_backend | }
app_backend | 2021-06-23T04:33:06.526Z ioredis:redis status[127.0.0.1:7001]: connecting -> close
app_backend | 2021-06-23T04:33:06.527Z ioredis:connection skip reconnecting because `retryStrategy` is not a function
app_backend | 2021-06-23T04:33:06.527Z ioredis:redis status[127.0.0.1:7001]: close -> end
app_backend | 2021-06-23T04:33:06.528Z ioredis:cluster:connectionPool Remove 127.0.0.1:7001 from the pool
app_backend | 2021-06-23T04:33:06.528Z ioredis:cluster Ready check failed (Error: Connection is closed.
app_backend |     at close (/usr/src/node_modules/ioredis/built/redis/event_handler.js:183:25)
app_backend |     at Socket.<anonymous> (/usr/src/node_modules/ioredis/built/redis/event_handler.js:154:20)
app_backend |     at Object.onceWrapper (events.js:482:26)
app_backend |     at Socket.emit (events.js:375:28)
app_backend |     at Socket.emit (domain.js:470:12)
app_backend |     at TCP.<anonymous> (net.js:675:12)). Reconnecting...
app_backend | 2021-06-23T04:33:06.530Z ioredis:cluster status: connect -> disconnecting
app_backend | 2021-06-23T04:33:06.530Z ioredis:redis status[127.0.0.1:7002 (ioredis-cluster(subscriber))]: wait -> close
app_backend | 2021-06-23T04:33:06.531Z ioredis:connection skip reconnecting since the connection is manually closed.
app_backend | 2021-06-23T04:33:06.531Z ioredis:redis status[127.0.0.1:7002 (ioredis-cluster(subscriber))]: close -> end
app_backend | 2021-06-23T04:33:06.531Z ioredis:cluster:subscriber stopped
app_backend | 2021-06-23T04:33:06.531Z ioredis:cluster:connectionPool Reset with []
app_backend | 2021-06-23T04:33:06.531Z ioredis:cluster:connectionPool Disconnect 127.0.0.1:7000 because the node does not hold any slot
app_backend | 2021-06-23T04:33:06.531Z ioredis:redis status[127.0.0.1:7000]: wait -> close
app_backend | 2021-06-23T04:33:06.531Z ioredis:connection skip reconnecting since the connection is manually closed.
app_backend | 2021-06-23T04:33:06.532Z ioredis:redis status[127.0.0.1:7000]: close -> end
app_backend | 2021-06-23T04:33:06.532Z ioredis:cluster:connectionPool Remove 127.0.0.1:7000 from the pool
app_backend | 2021-06-23T04:33:06.532Z ioredis:cluster:connectionPool Disconnect 127.0.0.1:7004 because the node does not hold any slot
app_backend | 2021-06-23T04:33:06.532Z ioredis:redis status[127.0.0.1:7004]: wait -> close
app_backend | 2021-06-23T04:33:06.532Z ioredis:connection skip reconnecting since the connection is manually closed.
app_backend | 2021-06-23T04:33:06.532Z ioredis:redis status[127.0.0.1:7004]: close -> end
app_backend | 2021-06-23T04:33:06.532Z ioredis:cluster:connectionPool Remove 127.0.0.1:7004 from the pool
app_backend | 2021-06-23T04:33:06.533Z ioredis:cluster:connectionPool Disconnect 127.0.0.1:7005 because the node does not hold any slot
app_backend | 2021-06-23T04:33:06.533Z ioredis:redis status[127.0.0.1:7005]: wait -> close
app_backend | 2021-06-23T04:33:06.533Z ioredis:connection skip reconnecting since the connection is manually closed.
app_backend | 2021-06-23T04:33:06.533Z ioredis:redis status[127.0.0.1:7005]: close -> end
app_backend | 2021-06-23T04:33:06.533Z ioredis:cluster:connectionPool Remove 127.0.0.1:7005 from the pool
app_backend | 2021-06-23T04:33:06.533Z ioredis:cluster:connectionPool Disconnect 127.0.0.1:7002 because the node does not hold any slot
app_backend | 2021-06-23T04:33:06.534Z ioredis:redis status[127.0.0.1:7002]: wait -> close
app_backend | 2021-06-23T04:33:06.534Z ioredis:connection skip reconnecting since the connection is manually closed.
app_backend | 2021-06-23T04:33:06.534Z ioredis:redis status[127.0.0.1:7002]: close -> end
app_backend | 2021-06-23T04:33:06.535Z ioredis:cluster:connectionPool Remove 127.0.0.1:7002 from the pool
app_backend | 2021-06-23T04:33:06.535Z ioredis:cluster:connectionPool Disconnect 127.0.0.1:7003 because the node does not hold any slot
app_backend | 2021-06-23T04:33:06.535Z ioredis:redis status[127.0.0.1:7003]: wait -> close
app_backend | 2021-06-23T04:33:06.535Z ioredis:connection skip reconnecting since the connection is manually closed.
app_backend | 2021-06-23T04:33:06.536Z ioredis:redis status[127.0.0.1:7003]: close -> end
app_backend | 2021-06-23T04:33:06.536Z ioredis:cluster:connectionPool Remove 127.0.0.1:7003 from the pool
app_backend | 2021-06-23T04:33:06.537Z ioredis:cluster status: disconnecting -> close
app_backend | 2021-06-23T04:33:06.537Z ioredis:cluster status: close -> close
app_backend | 2021-06-23T04:33:06.537Z ioredis:cluster status: close -> close
app_backend | 2021-06-23T04:33:06.537Z ioredis:cluster status: close -> close
app_backend | 2021-06-23T04:33:06.537Z ioredis:cluster status: close -> close
app_backend | 2021-06-23T04:33:06.538Z ioredis:cluster status: close -> end
app_backend | 2021-06-23T04:33:06.539Z ioredis:redis status[192.168.48.3:7000 (ioredis-cluster(refresher))]: ready -> close
app_backend | 2021-06-23T04:33:06.539Z ioredis:connection skip reconnecting since the connection is manually closed.
app_backend | 2021-06-23T04:33:06.539Z ioredis:redis status[192.168.48.3:7000 (ioredis-cluster(refresher))]: close -> end:
```

A few things that look suspicious to me are the `ioredis:cluster:connectionPool Reset` and `ioredis:cluster:connectionPool Disconnect 192.168.48.3:7000 because the node does not hold any slot` and `Error: connect ECONNREFUSED 127.0.0.1:7001` - but to be fair, I really don't have enough experience to know entirely what I'm looking at.

Lastly, in case it's useful, here's the output from the `redis-cluster` container itself:

```
-- IP Before trim: '0.0.0.0'
-- IP Before split: '0.0.0.0'
-- IP After trim: '0.0.0.0'
bash: warning: setlocale: LC_ALL: cannot change locale (en_US.UTF-8)
Using redis-cli to create the cluster
>>> Performing hash slots allocation on 6 nodes...
Master[0] -> Slots 0 - 5460
Master[1] -> Slots 5461 - 10922
Master[2] -> Slots 10923 - 16383
Adding replica 0.0.0.0:7004 to 0.0.0.0:7000
Adding replica 0.0.0.0:7005 to 0.0.0.0:7001
Adding replica 0.0.0.0:7003 to 0.0.0.0:7002
>>> Trying to optimize slaves allocation for anti-affinity
[WARNING] Some slaves are in the same host as their master
M: 261bcee7f8688a82d31a1b6396e5aa204073d1ec 0.0.0.0:7000
slots:[0-5460] (5461 slots) master
M: 1e00a469f940890d5e1ea8de0353c82e477cf4bf 0.0.0.0:7001
slots:[5461-10922] (5462 slots) master
M: 90cbfa7af8b322b86a2d716805f1ec3f0c383dfa 0.0.0.0:7002
slots:[10923-16383] (5461 slots) master
S: 19274fcbaaf2efa3b915bf96870cd2eb088d0dda 0.0.0.0:7003
replicates 90cbfa7af8b322b86a2d716805f1ec3f0c383dfa
S: 3a7b7c14a335245c3715205edb9b4cdd6f1b5af4 0.0.0.0:7004
replicates 261bcee7f8688a82d31a1b6396e5aa204073d1ec
S: d93f53bc66735830fee330968b82ed642b96b270 0.0.0.0:7005
replicates 1e00a469f940890d5e1ea8de0353c82e477cf4bf
Can I set the above configuration? (type 'yes' to accept): >>> Nodes configuration updated
>>> Assign a different config epoch to each node
>>> Sending CLUSTER MEET messages to join the cluster
Waiting for the cluster to join
.
>>> Performing Cluster Check (using node 0.0.0.0:7000)
M: 261bcee7f8688a82d31a1b6396e5aa204073d1ec 0.0.0.0:7000
slots:[0-5460] (5461 slots) master
1 additional replica(s)
S: 3a7b7c14a335245c3715205edb9b4cdd6f1b5af4 127.0.0.1:7004
slots: (0 slots) slave
replicates 261bcee7f8688a82d31a1b6396e5aa204073d1ec
S: d93f53bc66735830fee330968b82ed642b96b270 127.0.0.1:7005
slots: (0 slots) slave
replicates 1e00a469f940890d5e1ea8de0353c82e477cf4bf
M: 1e00a469f940890d5e1ea8de0353c82e477cf4bf 127.0.0.1:7001
slots:[5461-10922] (5462 slots) master
1 additional replica(s)
S: 19274fcbaaf2efa3b915bf96870cd2eb088d0dda 127.0.0.1:7003
slots: (0 slots) slave
replicates 90cbfa7af8b322b86a2d716805f1ec3f0c383dfa
M: 90cbfa7af8b322b86a2d716805f1ec3f0c383dfa 127.0.0.1:7002
slots:[10923-16383] (5461 slots) master
1 additional replica(s)
[OK] All nodes agree about slots configuration.
>>> Check for open slots...
>>> Check slots coverage...
[OK] All 16384 slots covered.
==> /var/log/supervisor/redis-1.log <==
43:M 23 Jun 2021 04:32:44.200 * Ready to accept connections
43:M 23 Jun 2021 04:32:46.211 # configEpoch set to 1 via CLUSTER SET-CONFIG-EPOCH
43:M 23 Jun 2021 04:32:46.300 # IP address for this node updated to 127.0.0.1
43:M 23 Jun 2021 04:32:48.185 * Replica 127.0.0.1:7004 asks for synchronization
43:M 23 Jun 2021 04:32:48.185 * Partial resynchronization not accepted: Replication ID mismatch (Replica asked for 'a38359fb1087ee7e0b0fa407caf77be85473d2ce', my replication IDs are '911b937c2a3624b9f15d2a20952d60049eeb8d3' and '0000000000000000000000000000000000000000')
43:M 23 Jun 2021 04:32:48.185 * Replication backlog created, my new replication IDs are '02f791228377e9b214608bc7f0c3b030aca93bc5' and'0000000000000000000000000000000000000000'
43:M 23 Jun 2021 04:32:48.185 * Starting BGSAVE forSYNC with target: disk
43:M 23 Jun 2021 04:32:48.186 * Background saving tarted by pid 76
76:C 23 Jun 2021 04:32:48.188 * DB saved on dis
76:C 23 Jun 2021 04:32:48.188 * RDB: 0 MB of emory used by copy-on-write
==> /var/log/supervisor/redis-2.log <==
45:M 23 Jun 2021 04:32:46.212 # configEpoch set to 2 via CLUSTER SET-CONFIG-EPOCH
45:M 23 Jun 2021 04:32:46.405 # IP address for this node updated to 127.0.0.1
45:M 23 Jun 2021 04:32:46.405 # Address updated for node 261bcee7f8688a82d31a1b6396e5aa204073d1ec, now 127.0.0.1:7000
45:M 23 Jun 2021 04:32:48.192 * Replica 127.0.0.1:7005 asks for synchronization
45:M 23 Jun 2021 04:32:48.192 * Partial resynchroniztion not accepted: Replication ID mismatch (Replica asked for '58d158b99c5b3b4622de156bc111d9bbb84f21da', myreplication IDs are '76fe7a8979702156824dd700faf9effb300020ba' and '0000000000000000000000000000000000000000')
45M 23 Jun 2021 04:32:48.192 * Replication backlog created, my new replication IDs are'f7af9689b55b60b19ff7bdc04c477562c3faf5d7' and '0000000000000000000000000000000000000000'
45:M 23 Jun 2021 04:32:48.192 * Starting BGSAVE for SYNC with target: disk
45:M 23 Jun 2021 04:32:48.193 * Background saving started by pid 77
77:C 23 Jun 2021 04:32:48.195 * DB saved on disk
77:C 23 Jun 2021 04:32:48.195 * RDB: 0 MB of memory used by copy-on-write

==> /var/log/supervisor/redis-3.log <==
44:M 23 Jun 2021 04:32:46.212 # configEpoch set to 3 via CLUSTER SET-CONFIG-EPOCH
44:M 23 Jun 2021 04:32:46.405 # IP address for this node updated to 127.0.0.1
44:M 23 Jun 2021 04:32:46.405 # Address updated for node 261bcee7f8688a82d31a1b6396e5aa204073d1ec, now 127.0.0.1:7000
44:M 23 Jun 2021 04:32:48.185 * Replica 127.0.0.1:7003 asks for synchronization
44:M 23 Jun 2021 04:32:48.185 * Partial resynchroniztion not accepted: Replication ID mismatch (Replica asked for '9ca7c4cb1d40d57b4c2f6d9147d487a395cca0e2', myreplication IDs are '336554a77ed21bc9900cbc251d3e9485252c4524' and '0000000000000000000000000000000000000000')
44M 23 Jun 2021 04:32:48.185 * Replication backlog created, my new replication IDs are'a67956d505d5ade21f6746c0ec489bc14e0066e8' and '0000000000000000000000000000000000000000'
44:M 23 Jun 2021 04:32:48.185 * Starting BGSAVE for SYNC with target: disk
44:M 23 Jun 2021 04:32:48.185 * Background saving started by pid 75
75:C 23 Jun 2021 04:32:48.188 * DB saved on disk
75:C 23 Jun 2021 04:32:48.188 * RDB: 0 MB of memory used by copy-on-write

==> /var/log/supervisor/redis-4.log <==
41:M 23 Jun 2021 04:32:46.304 # Address updated for node 261bcee7f8688a82d31a1b6396e5aa204073d1ec, now 127.0.0.1:7000
41: 23 Jun 2021 04:32:48.184 * Before turning into a replica, using my own master parameters to synthesize a cached master: I maybe able to synchronize with the new master with just a partial transfer.
41:S 23 Jun 2021 04:32:48.184 * Connecting to MASTER 127.0.0.1:7002
41:S 23 Jun 2021 04:32:48.184 * MASTER <-> REPLICA sync started
41:S 23 Jun 2021 04:32:48.184 # Cluster state changed: ok
41:S 23 Jun 2021 04:32:48.184 * Non blocking connect for SYNC fired the event.
41:S 23 Jun 2021 04:32:48.184 * Master replied to PING, replication can continue...
41:S 23 Jun 2021 04:32:48.185 * Trying a partial resynchronization (request 9ca7c4cb1d40d57b4c2f6d9147d487a395cca0e2:1).
41:S 23 Jun 2021 04:32:48.186 * Full resync from master: a67956d505d5ade21f6746c0ec489bc14e0066e8:0
41:S 23 Jun 2021 04:32:48.186 * Discarding previously cached master state.

==> /var/log/supervisor/redis-5.log <==
40:M 23 Jun 2021 04:32:46.304 # Address updated for node 261bcee7f8688a82d31a1b6396e5aa204073d1ec, now 127.0.0.1:7000
40: 23 Jun 2021 04:32:48.184 * Before turning into a replica, using my own master parameters to synthesize a cached master: I maybe able to synchronize with the new master with just a partial transfer.
40:S 23 Jun 2021 04:32:48.184 * Connecting to MASTER 127.0.0.1:7000
40:S 23 Jun 2021 04:32:48.184 * MASTER <-> REPLICA sync started
40:S 23 Jun 2021 04:32:48.184 # Cluster state changed: ok
40:S 23 Jun 2021 04:32:48.184 * Non blocking connect for SYNC fired the event.
40:S 23 Jun 2021 04:32:48.185 * Master replied to PING, replication can continue...
40:S 23 Jun 2021 04:32:48.185 * Trying a partial resynchronization (request a38359fb1087ee7e0b0fa407caf77be85473d2ce:1).
40:S 23 Jun 2021 04:32:48.186 * Full resync from master: 02f791228377e9b214608bc7f0c3b030aca93bc5:0
40:S 23 Jun 2021 04:32:48.186 * Discarding previously cached master state.

==> /var/log/supervisor/redis-6.log <==
42:M 23 Jun 2021 04:32:46.304 # Address updated for node 261bcee7f8688a82d31a1b6396e5aa204073d1ec, now 127.0.0.1:7000
42:S 23 Jun 2021 04:32:48.185 * Before turning into a replica, using my own master parameters to synthesize a cached master: I may be able to synchronize with the new master with just a partial transfer.
42:S 23 Jun 2021 04:32:48.185 * Connecting to MASTER 127.0.0.1:7001
42:S 23 Jun 2021 04:32:48.185 * MASTER <-> REPLICA sync started
42:S 23 Jun 2021 04:32:48.185 # Cluster state changed: ok
42:S 23 Jun 2021 04:32:48.185 * Non blocking connect for SYNC fired the event.
42:S 23 Jun 2021 04:32:48.185 * Master replied to PING, replication can continue...
42:S 23 Jun 2021 04:32:48.187 * Trying a partial resynchronization (request 58d158b99c5b3b4622de156bc111d9bbb84f21da:1).
42:S 23 Jun 2021 04:32:48.193 * Full resync from master: f7af9689b55b60b19ff7bdc04c477562c3faf5d7:0
42:S 23 Jun 2021 04:32:48.193 * Discarding previously cached master state.

==> /var/log/supervisor/redis-1.log <==
43:M 23 Jun 2021 04:32:48.275 * Background saving terminated with success
43:M 23 Jun 2021 04:32:48.276 * Synchronization with replica 127.0.0.1:7004 succeeded
==> /var/log/supervisor/redis-2.log <==
45:M 23 Jun 2021 04:32:48.277 * Background saving terminated with success
45:M 23 Jun 2021 04:32:48.278 * Synchronization with replica 127.0.0.1:7005 succeeded

==> /var/log/supervisor/redis-3.log <==
44:M 23 Jun 2021 04:32:48.275 * Background saving terminated with success
44:M 23 Jun 2021 04:32:48.275 * Synchronization with replica 127.0.0.1:7003 succeeded
==> /var/log/supervisor/redis-4.log <==
41:S 23 Jun 2021 04:32:48.276 * MASTER <-> REPLICA sync: receiving 175 bytes from master to disk
41:S 23 Jun 2021 04:32:48.276 * MASTER <-> REPLICA sync: Flushing old data
41:S 23 Jun 2021 04:32:48.276 * MASTER <-> REPLICA sync: Loading DB in memory
41:S 23 Jun 2021 04:32:48.279 * Loading RDB produced by version 6.2.0
41:S 23 Jun 2021 04:32:48.279 * RDB age 0 seconds
41:S 23 Jun 2021 04:32:48.279 * RDB memory usage when created 2.60 Mb
41:S 23 Jun 2021 04:32:48.279 * MASTER <-> REPLICA sync: Finished with success
41:S 23 Jun 2021 04:32:48.280 * Background append only file rewriting started by pid 79
41:S 23 Jun 2021 04:32:48.308 * AOF rewrite child asks to stop sending diffs.
79:C 23 Jun 2021 04:32:48.309 * Parent agreed to stop sending diffs. Finalizing AOF...
79:C 23 Jun 2021 04:32:48.309 * Concatenating 0.00 MB of AOF diff received from parent.
79:C 23 Jun 2021 04:32:48.309 * SYNC append only file rewrite performed
79:C 23 Jun 2021 04:32:48.310 * AOF rewrite: 0 MB of memory used by copy-on-write
41:S 23 Jun 2021 04:32:48.374 * Background AOF rewrite terminated with success
41:S 23 Jun 2021 04:32:48.374 * Residual parent diff successfully flushed to the rewritten AOF (0.00 MB)
41:S 23 Jun 2021 04:32:48.374 * Background AOF rewrite finished successfully

==> /var/log/supervisor/redis-5.log <==
40:S 23 Jun 2021 04:32:48.276 * MASTER <-> REPLICA sync: receiving 175 bytes from master to disk
40:S 23 Jun 2021 04:32:48.276 * MASTER <-> REPLICA sync: Flushing old data
40:S 23 Jun 2021 04:32:48.276 * MASTER <-> REPLICA sync: Loading DB in memory
40:S 23 Jun 2021 04:32:48.281 * Loading RDB produced by version 6.2.0
40:S 23 Jun 2021 04:32:48.281 * RDB age 0 seconds
40:S 23 Jun 2021 04:32:48.281 * RDB memory usage when created 2.59 Mb
40:S 23 Jun 2021 04:32:48.281 * MASTER <-> REPLICA sync: Finished with success
40:S 23 Jun 2021 04:32:48.282 * Background append only file rewriting started by pid 80
40:S 23 Jun 2021 04:32:48.315 * AOF rewrite child asks to stop sending diffs.
80:C 23 Jun 2021 04:32:48.315 * Parent agreed to stop sending diffs. Finalizing AOF...
80:C 23 Jun 2021 04:32:48.315 * Concatenating 0.00 MB of AOF diff received from parent.
80:C 23 Jun 2021 04:32:48.315 * SYNC append only file rewrite performed
80:C 23 Jun 2021 04:32:48.316 * AOF rewrite: 0 MB of memory used by copy-on-write
40:S 23 Jun 2021 04:32:48.370 * Background AOF rewrite terminated with success
40:S 23 Jun 2021 04:32:48.370 * Residual parent diff successfully flushed to the rewritten AOF (0.00 MB)
40:S 23 Jun 2021 04:32:48.371 * Background AOF rewrite finished successfully
==> /var/log/supervisor/redis-6.log <==
42:S 23 Jun 2021 04:32:48.278 * MASTER <-> REPLICA sync: receiving 175 bytes from master to disk
42:S 23 Jun 2021 04:32:48.278 * MASTER <-> REPLICA sync: Flushing old data
42:S 23 Jun 2021 04:32:48.278 * MASTER <-> REPLICA sync: Loading DB in memory
42:S 23 Jun 2021 04:32:48.282 * Loading RDB produced by version 6.2.0
42:S 23 Jun 2021 04:32:48.282 * RDB age 0 seconds
42:S 23 Jun 2021 04:32:48.282 * RDB memory usage when created 2.56 Mb
42:S 23 Jun 2021 04:32:48.283 * MASTER <-> REPLICA sync: Finished with success
42:S 23 Jun 2021 04:32:48.283 * Background append only file rewriting started by pid 81
42:S 23 Jun 2021 04:32:48.312 * AOF rewrite child asks to stop sending diffs.
81:C 23 Jun 2021 04:32:48.312 * Parent agreed to stop sending diffs. Finalizing AOF...
81:C 23 Jun 2021 04:32:48.312 * Concatenating 0.00 MB of AOF diff received from parent.
81:C 23 Jun 2021 04:32:48.313 * SYNC append only file rewrite performed
81:C 23 Jun 2021 04:32:48.314 * AOF rewrite: 0 MB of memory used by copy-on-write
42:S 23 Jun 2021 04:32:48.374 * Background AOF rewrite terminated with success
42:S 23 Jun 2021 04:32:48.374 * Residual parent diff successfully flushed to the rewritten AOF (0.00 MB)
42:S 23 Jun 2021 04:32:48.374 * Background AOF rewrite finished successfully

==> /var/log/supervisor/redis-1.log <==
43:M 23 Jun 2021 04:32:51.282 # Cluster state changed: ok

==> /var/log/supervisor/redis-2.log <==
45:M 23 Jun 2021 04:32:51.283 # Cluster state changed: ok

==> /var/log/supervisor/redis-3.log <==
44:M 23 Jun 2021 04:32:51.282 # Cluster state changed: ok
```

Any help would be much appreciated!

**Comment by Grokzen** [comment]
@tjosephs Hi and thank you for using the discussions sections for this.

So first i do not know how IORedis works so i can't speak for that part. I do know exactly what the error really is but i can't say for sure what you need to do to solve it fully.

So this problem is really a common problem when doing redis cluster nodes inside docker specially and i am sure that running docker on mac as docker on mac is not always the exact same and works perfect as you want, it will cause issues like this.

So your error. In redis cluster there is basically two different IP:s that if they are not in sync and reachable, they will not work out properly. One IP is the IP that you use with your client lib, IORedis in this case, to do the first initial connection to any cluster node to let the client configure itself and to fetch where all other nodes are. In your case you set this to you 192.x.x.x address. This address must be reachable from where you execute your client code and port 7000-7005 must be open on that IP as well.

The other side of this is what IP address that each redis node uses internally to talk between each node. This node discovery process is one step that is done in the Docker container entrypoints and since all nodes lives inside the same docker container, they can all talk over 127.0.0.1:<port> instead of using some other IP like the container IP or a public machine IP, redis will use the simplest mechanism for this.

So the issue here and with all clients is that after you have connected to a redis cluster and done your node discovery step, what you see in IORedis is that "all nodes is on 127.0.0.1" so the client will think that, "hey the node i want is on 127.0.0.1" and when you are using mac this is the main issue where when running docker it dont always udnerstand that 127.0.0.1 should look inside the docker container or to bind the ports to that outside the docker container. However what is supposed to happen is that when you specify IP=0.0.0.0 what should happen is that on the inside of the container it should default back to binding all the nodes to 0.0.0.0:<port> instead of 127.0.0.1 or if it was the other way around that docker should bind it to 0.0.0.0 on your host machine, i dont really remember and i do not have a mac machine to test this out myself on really.

So when this 127.0.0.1 address comes back to your client lib, the lib will bind each slot in the cluster to each IP the cluster says they live on. But if your client do not know how to route or get to the IP that redis says it should route to, then you get this issue and you get a failure in your client layer.

So in short, this is what the issue is and why it keeps happening, i can't however help so much on mac as i can't debug it myself and i have to have help from the community to test these things out for me.

**Reply by tjosephs** [reply]
Thank you so much for the timely response!!! That makes total sense of the issue. I had pretty much been able to figure out the two addresses situation and that 127.0.0.1 was internal to the container, but what I didn't realize was that my client was trying to then turn around and talk to localhost again from its own perspective.

This gives me some definite ideas to follow up on. I'll spend some more time with it and see where I can get and will then report back.

**Comment by tjosephs** [comment, marked as answer]
Figured it out. `IP=0.0.0.0` was doing the exact opposite of what we were wanting it to. It was causing the nodes inside the container to bind to `localhost`. That led to the problems you described above. As soon as I removed that line, the nodes bound to the external IP address on the container and I was able to address them from my application container.

Final `docker-compose.yml` looked like this:

```
redis:
    image: grokzen/redis-cluster:latest
    ports:
      - '7000-7005:7000-7005'
      - '5000-5010:5000-5010'
```

About as easy as it gets.

Now we're up and running! Container and app seem to be working well! We're about to try it in AWS with ElastiCache. Once I know that's working I'll post back here as well just to close the loop all the way.

**Reply by tjosephs** [reply]
And confirmed - the code we were running locally against your container worked perfectly once we got it in ElastiCache. 

Thanks for providing a great solution for local dev!
